File: lintcode/16_permutations-ii/permutations-ii.cpp
/*
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/permutations-ii
@Language: C++
@Datetime: 16-05-04 20:44
*/

class Solution {
public:
    /**
     * @param nums: A list of integers.
     * @return: A list of unique permutations.
     */
    vector<vector<int> > permuteUnique(vector<int> &nums) {
        vector<vector<int> > res;
        util(nums, nums.size()-1, nums.size(), res);
        return res;
    }
    
    void util(vector<int> &nums, int p, int n, vector<vector<int> > &res) {
        if (p == -1) {
            res.push_back(nums);
            return;
        }

        for (int i = p; i < n; ++i) {
            if (i != p && nums[p] == nums[i]) break;
            swap(nums[p], nums[i]);
            util(nums, p-1, n, res);
            swap(nums[p], nums[i]);
        }
    }
};

File: lintcode/16_permutations-ii/permutations-ii.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/permutations-ii
@Language: Python
@Datetime: 16-07-07 23:17
'''

class Solution:
    """
    @param nums: A list of integers.
    @return: A list of unique permutations.
    """
    def permuteUnique(self, nums):
        n = len(nums)
        res = []
        def helper(p):
            if p == -1:
                res.append(nums[:])
            else:
                for i in range(p, n):
                    if i > p and nums[i] == nums[p]:
                        break
                    nums[i], nums[p] = nums[p], nums[i]
                    helper(p-1)
                    nums[i], nums[p] = nums[p], nums[i]
        helper(n-1)
        return res